zircon-object/src/debuglog.rs

```diff
@@ -0,0 +1,219 @@
+use {
+    super::*,
+    crate::hal::{serial_write, timer_now},
+    crate::object::*,
+    alloc::sync::Arc,
+    serde::{Deserialize, Serialize},
+    spin::Mutex,
+};
+
+static DLOG: Mutex<DlogBuffer> = Mutex::new(DlogBuffer::new());
+
+#[cfg(not(test))]
+const DLOG_SIZE: usize = 128usize * 1024usize;
+#[cfg(test)]
+const DLOG_SIZE: usize = 2usize * 1024usize;
+const DLOG_MASK: usize = DLOG_SIZE - 1;
+const DLOG_MIN_RECORD: usize = 32usize;
+
+#[derive(Serialize, Deserialize)]
+#[repr(C)]
+struct DlogHeader {
+    pub header: u32,
+    pub datalen: u16,
+    pub flags: u16,
+    pub timestamp: u64,
+    pub pid: u64,
+    pub tid: u64,
+}
+
+pub struct DebugLog {
+    base: KObjectBase,
+    #[allow(dead_code)]
+    flags: u32,
+}
+
+impl_kobject!(DebugLog);
+
+impl DebugLog {
+    pub fn create(flags: u32) -> ZxResult<Arc<Self>> {
+        let dlog = Arc::new(DebugLog {
+            base: KObjectBase::new(),
+            flags,
+        });
+        Ok(dlog)
+    }
+
+    pub fn write(&self, flags: u32, data: &str) -> ZxResult<usize> {
+        let flags = flags | self.flags;
+        DLOG.lock().write(flags, data.as_bytes());
+        serial_write(data);
+        serial_write("\n");
+        Ok(0)
+    }
+}
+
+#[allow(dead_code)]
+struct DlogBuffer {
+    buf: [u8; DLOG_SIZE],
+    head: usize,
+    tail: usize,
+}
+
+impl DlogBuffer {
+    pub const fn new() -> Self {
+        DlogBuffer {
+            buf: [0u8; DLOG_SIZE],
+            head: 0usize,
+            tail: 0usize,
+        }
+    }
+
+    pub fn write(&mut self, flags: u32, data: &[u8]) {
+        let wire_size = DLOG_MIN_RECORD + ((data.len() + 3) & !3);
+        let header = (((DLOG_MIN_RECORD + data.len()) as u32 & 0xFFFu32) << 12)
+            | (wire_size as u32 & 0xFFFu32);
+        let serde_header = bincode::serialize(&DlogHeader {
+            header,
+            datalen: data.len() as u16,
+            flags: flags as u16,
+            timestamp: timer_now().as_nanos() as u64,
+            pid: 0u64,
+            tid: 0u64,
+        })
+        .unwrap();
+        let head = self.head;
+        while (head - self.tail) > (DLOG_SIZE - wire_size) {
+            let tail_index = self.tail & DLOG_MASK;
+            let header: u32 =
+                bincode::deserialize::<u32>(&self.buf[tail_index..tail_index + 4]).unwrap();
```

Comment:
`&[u8] === try_into ==> [u8; 4] === from_ne_bytes ==> u32`

```suggestion
                u32::from_ne_bytes(self.buf[tail_index..tail_index + 4].try_into().unwrap());
```